src/Lemon/Containers/Converters/Ncch2Binary.cs

```diff
@@ -0,0 +1,237 @@
+
+            writer.Stream.Position = 0x190;
+            WriteOffsetSizeAndData(writer, root.Children["sdk_info.txt"], source.Header.SdkInfoOffset);
+            WriteOffsetSizeAndData(writer, root.Children["logo.bin"], source.Header.LogoOffset);
+            WriteOffsetSizeAndData(writer, root.Children["system"], source.Header.SystemOffset, source.Header.SystemHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+            WriteOffsetSizeAndData(writer, root.Children["rom"], source.Header.RomOffset, source.Header.RomHashSize, true);
+
+            writer.Stream.Position += 0x4; // Reserved
+
+            WriteSHA256(writer, root.Children["system"], source.Header.SystemHashSize, true);
+            WriteSHA256(writer, root.Children["rom"], source.Header.RomHashSize, true);
+
+            // Write the full size of the binary (in units) in 0x104
+            writer.Stream.Position = 0x104;
+            writer.Write((int)writer.Stream.Length / NcchHeader.Unit);
+
+            return binary;
+        }
+
+        void WriteFile(DataWriter writer, Node file, bool isRequired = true)
+        {
+            if (file == null && isRequired) {
+                throw new System.IO.FileNotFoundException("Missing NCCH file");
+            } else if (file == null) {
+                return;
+            }
+
+            if (file.Format is not IBinary) {
+                throw new FormatException("Cannot write file as it is not binary");
+            }
+
+            file.Stream.WriteTo(writer.Stream);
+        }
+
+        void WriteOffsetSizeAndData(DataWriter writer, Node file, int offsetInUnits, int hashRegion = 0, bool hasHashRegion = false)
+        {
+            if (file != null) {
+                long position = writer.Stream.Position;
+                int offset = offsetInUnits * NcchHeader.Unit;
+                int fileLength = (int)file.Stream.Length;
+                int fileLengthInUnits = fileLength / NcchHeader.Unit;
+
+                bool usedChunk = false;
+                var fileRange = Enumerable.Range(offset, fileLength);
+                foreach (var usedUnit in usedUnits) {
+                    var usedRange = Enumerable.Range(usedUnit.Key, usedUnit.Value);
+
+                    if (fileRange.Intersect(usedRange).Any()) {
+                        usedChunk = true;
+                    }
+                }
+
+                if (writer.Stream.Length > offset && usedChunk) {
+                    offset = (int)Math.Ceiling((double)writer.Stream.Length / NcchHeader.Unit) * NcchHeader.Unit;
+                    offsetInUnits = offset / NcchHeader.Unit;
+                }
+
+                if (writer.Stream.Length < offset) {
+                    long tempPosition = writer.Stream.Position;
+
+                    writer.Stream.Position = writer.Stream.Length;
+                    writer.Write(new byte[offset - writer.Stream.Length]);
+
+                    writer.Stream.Position = tempPosition;
+                }
```

Comment:
I think we needed to do this because we are not writing the padding after each file. I think we won't needed once you add the line from the comment below.

Reply:
That's partially the reason, but it's also done for generating the same stream while testing and if the content has changed then it calculates a new position at the end of the file. But yeah, it could use some padding.